### PR TITLE
Add support for --coverage to flutter test

### LINKED ISSUE #2342: Would be nice to produce code coverage from `flutter test`
Should be possible to add a --coverage option to flutter test and integrate with package:coverage.
https://github.com/dart-lang/coverage

### PR DESCRIPTION
**Pull request by abarth**
We need https://github.com/dart-lang/tools/issues/561 to be fixed before
this will be useful.

Fixes #2342

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @cbracken

**Comment by cbracken** [comment]
Heads up: dart-lang/coverage#100 is now fixed.

File: packages/flutter_tools/lib/src/test/coverage_collector.dart [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  List<Future<Null>> _jobs = <Future<Null>>[];
+  Map<String, dynamic> _globalHitmap;
+
+  Future<String> finalizeCoverage() async {
+    assert(enabled);
+    await finishPendingJobs();
+    printTrace('formating coverage data');
+    // TODO(abarth): Use PackageMap.globalPackagesPath once
+    // https://github.com/dart-lang/coverage/issues/100 is fixed.
+    Resolver resolver = new Resolver(packageRoot: path.absolute('packages'));
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/test/coverage_collector.dart]: New param in 0.7.8 is `packagesPath` (String path to .packages file)

**Comment by eseidelGoogle** [comment]
It's clearly dangerous putting you two in a room together. :)

So excited to try this out!

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 2]
```diff
@@ -105,8 +112,16 @@ void main() {
 }
 ''');
 
+    int observatoryPort;
```
Comment (eseidelGoogle) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: My understanding was the dart vm would pick an open port if you passed --observatory-port=0, and then we had code to read said port from the logs, etc.  Maybe that won't work through our package:test integration?  Devon might know.

Comment (abarth) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Yeah, I didn't want to mess with reading the port back from the logs.  Seemed easier to push the port number in via a command-line parameter.

Comment (devoncarew) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: We do have code in `os.dart` - `findAvailablePort()` - that tries to find an available local port. I'm not sure it's more reliable than the random generator below however -

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 3]
```diff
@@ -13,12 +13,21 @@ import '../cache.dart';
...
 
 class TestCommand extends FlutterCommand {
   TestCommand() {
     usesPubOption();
+    argParser.addFlag('coverage',
+      defaultsTo: false,
+      help: 'Whether to collect coverage information.'
+    );
+    argParser.addOption('coverage-path',
+      defaultsTo: 'coverage/lcov.info',
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/test.dart]: what is this relative to?

Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/test.dart]: if it's relative to the current working directory, then having a directory as part of the default seems weird.

Comment (abarth) [inline packages/flutter_tools/lib/src/commands/test.dart]: It's relative to the current working directory.

It is strange to having a directory in the default, but that's the norm for lcov and what several automated tools expect you to do.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: Done.

**Comment by abarth** [comment]
This is ready for review now.

**Comment by Hixie** [comment]
LGTM

**Pull request merged**